addons/mail/models/mail_message.py

```diff
@@ -383,226 +383,187 @@ def check_access_rule(self, operation):
-
-        messages_to_check = set(messages_to_check).difference(set(notified_ids))
-        if not messages_to_check:
-            return
+        if operation in ('read', 'write'):
+            for mid, message in list(messages_to_check.items()):
+                if message.get('notified'):
+                    messages_to_check.pop(mid)
+            if not messages_to_check:
+                return result
 
         # CRUD: Access rights related to the document
-        document_related_ids = []
-        document_related_candidate_ids = [
-            mid for mid, message in message_values.items()
+        # {document_model_name: {document_id: message_ids}}
+        model_docid_msgids = defaultdict(lambda: defaultdict(list))
+        for mid, message in messages_to_check.items():
             if (message.get('model') and message.get('res_id') and
-                message.get('message_type') != 'user_notification')
-        ]
-        model_record_ids = _generate_model_record_ids(message_values, document_related_candidate_ids)
-        for model, doc_ids in model_record_ids.items():
-            DocumentModel = self.env[model]
-            if hasattr(DocumentModel, '_get_mail_message_access'):
-                check_operation = DocumentModel._get_mail_message_access(doc_ids, operation)  ## why not giving model here?
-            else:
-                check_operation = self.env['mail.thread']._get_mail_message_access(doc_ids, operation, model_name=model)
-            records = DocumentModel.browse(doc_ids)
-            records.check_access_rights(check_operation)
-            mids = records.browse(doc_ids)._filter_access_rules(check_operation)
-            document_related_ids += [
-                mid for mid, message in message_values.items()
-                if (
-                    message.get('model') == model and
-                    message.get('res_id') in mids.ids and
-                    message.get('message_type') != 'user_notification'
-                )
-            ]
+                    message.get('message_type') != 'user_notification'):
+                model_docid_msgids[message['model']][message['res_id']].append(mid)
 
-        messages_to_check = messages_to_check.difference(set(document_related_ids))
+        for model, docid_msgids in model_docid_msgids.items():
+            documents = self.env[model].browse(docid_msgids)
+            if hasattr(documents, '_get_mail_message_access'):
+                doc_operation = documents._get_mail_message_access(docid_msgids, operation)  # why not giving model here?
+            else:
+                doc_operation = self.env['mail.thread']._get_mail_message_access(docid_msgids, operation, model_name=model)
+            doc_result = documents._check_access(doc_operation)
+            forbidden_doc_ids = set(doc_result[0]) if doc_result else set()
+            for doc_id, msg_ids in docid_msgids.items():
+                if doc_id not in forbidden_doc_ids:
```

Comment:
`doc_id` is a int but `forbidden_doc_ids` is actually a set of recordset, right ?